src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
+        static final Object REPORT_ERROR = new Object();
+        /** Fast, unbounded emission complete. */
+        static final Object REPORT_COMPLETED = new Object();
+        
+        public MergeSubscriber(Subscriber<? super T> child, boolean delayErrors, int maxConcurrent) {
+            this.child = child;
+            this.delayErrors = delayErrors;
+            this.maxConcurrent = maxConcurrent;
+            this.nl = NotificationLite.instance();
+            this.innerGuard = new Object();
+            this.innerSubscribers = EMPTY;
+            this.request(0);
+        }
+        
+        Queue<Throwable> getOrCreateErrorQueue() {
+            ConcurrentLinkedQueue<Throwable> q = errors;
+            if (q == null) {
+                synchronized (this) {
+                    q = errors;
+                    if (q == null) {
+                        q = new ConcurrentLinkedQueue<Throwable>();
+                        errors = q;
+                    }
+                }
             }
+            return q;
         }
-
-        private void handleScalarSynchronousObservableWithoutRequestLimits(ScalarSynchronousObservable<? extends T> t) {
-            T value = t.get();
-            if (getEmitLock()) {
-                boolean moreToDrain;
-                try {
-                    actual.onNext(value);
-                } finally {
-                    moreToDrain = releaseEmitLock();
+        CompositeSubscription getOrCreateComposite() {
+            CompositeSubscription c = subscriptions;
+            if (c == null) {
+                boolean shouldAdd = false;
+                synchronized (this) {
+                    c = subscriptions;
+                    if (c == null) {
+                        c = new CompositeSubscription();
+                        subscriptions = c;
+                        shouldAdd = true;
+                    }
                 }
-                if (moreToDrain) {
-                    drainQueuesIfNeeded();
+                if (shouldAdd) {
+                    add(c);
                 }
-                request(1);
+            }
+            return c;
+        }
+        
+        /** 
+         * In case the downstream requests less than Long.MAX_VALUE initially, do the standard
+         * queued processing.
+         */
+        void requestDefault() {
+            request(Math.min(maxConcurrent, RxRingBuffer.SIZE));
+        }
+        /**
+         * In case the downstream requests Long.MAX_VALUE, perform a simplified merging without
+         * using queues at all.
+         */
+        void requestAll() {
+            FAST_WIP.lazySet(this, 1);
+            fastMode = true;
+            request(maxConcurrent == Integer.MAX_VALUE ? Long.MAX_VALUE : maxConcurrent);
+        }
+        
+        @Override
+        public void onNext(Observable<? extends T> t) {
+            if (t == null) {
                 return;
+            }
+            if (fastMode) {
```

Comment:
Volatile reads are cheap and since it won't change ever after, it will stay in L1. If I do a plain+volatile+plain read (similar to a double-checked locking) that plain variable is not guaranteed to stay in a register and will be promoted to L1 as well, so no win there.

The other concern that the JIT won't be able to do some optimizations due to its barrier nature is not an issue here because the subsequent operations do atomics and `synchronized` which have the same effect.